content/docs/using-pulumi/debugging/_index.md

```diff
@@ -0,0 +1,295 @@
+
+Depending on your setup, you may need to attach multiple times until the right node process is found. For example, you may be running Node.js policy packs that will hit the debugger as well.
+
+Now you can step through the program and inspect variables. Once completed, resume the execution and let the Pulumi CLI complete.
+
+## Debugging Python programs
+
+This guide uses [`debugpy`](https://github.com/microsoft/debugpy) to debug Pulumi programs.
+
+You need to install the `debugpy` package to your Pulumi program by adding it to `requirements.txt`:
+
+```txt
+pulumi>=3.0.0,<4.0.0
+debugpy>=1.0.0
+```
+
+Then, prepend your Pulumi program with the code to wait for a debugger to attach:
+
+```py
+import debugpy
+
+debugpy.listen(("localhost", 5678))
+debugpy.wait_for_client()
+
+# ... Your normal Pulumi code follows here
+```
+
+Run `pulumi up` the usual way:
+
+```bash
+pulumi up
+Previewing update (dev)
+
+ Type                 Name  Plan   Info
+ pulumi:pulumi:Stack  dev          0.00s - Note: Debugging will proceed.
+```
+
+Set a breakpoint in your Pulumi code.
+
+Navigate to your IDE and attach the debugger to the process with the `python` runtime. For example, in VS Code, you can define the following `launch.json` file:
+
+```json
+{
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "name": "Python: Local Attach",
+            "type": "debugpy",
+            "request": "attach",
+            "connect": {
+                "host": "localhost",
+                "port": 5678
+            },
+            "pathMappings": [
+                {
+                    "localRoot": "${workspaceFolder}",
+                    "remoteRoot": "."
+                }
+            ],
+            "justMyCode": true
+        }
+    ]
+}
+```
+
+navigate to the `Run and Debug` menu on the left pane, and then click the Start Debugging (F5) button with the `Python: Local Attach` configuration.
+
+Now you can step through the program and inspect variables. Once completed, resume the execution and let the Pulumi CLI complete.
+
+## Debugging .NET programs
+
+Prepend your Pulumi program with the code to wait for a debugger to attach:
+
+```cs
+using System.Diagnostics;
+
+return await Pulumi.Deployment.RunAsync(async () =>
+{
+    while (!Debugger.IsAttached)
+    {
+        await System.Threading.Tasks.Task.Delay(1000);
+    }
+    Debugger.Break();
+
+    // ... Your normal Pulumi code follows here
+});
+```
+
+Run `pulumi up` the usual way:
+
+```bash
+pulumi up
+Previewing update (dev)
+
+ Type                 Name          Plan   Info
+ pulumi:pulumi:Stack  play-cs-dev          'dotnet build -nologo .' completed successfully
+```
+
+Navigate to your IDE and attach the debugger to the process with the name of your .NET project. For example, in VS Code, you can define the following `launch.json` file:
```

Comment:
Any reason not to mention setting a breakpoint here also, as we do in the other sections?

Reply:
Because `Debugger.Break();` will break for the user. However, maybe it makes sense to unify this with other sections, remove `Debugger.Break();` and ask the user to set a breakpoint. I will do that.